### PR TITLE
[WIP] Feature/oauth/interceptor

### PR DESCRIPTION
**Pull request by SEQUOIIA**
**NOT TO BE MERGED, MERELY FOR DISCUSSION AND FEEDBACK**

Current state:

- Intercepts requests on endpoints deemed "protected"
- Upon interception, will check if user is logged in with msal, if not, logs in the user.
- Acquires needed access token using msal(their built-in caching seems to be broken, not sure if our fault or theirs, therefore I've added a WIP cache that resets on full page reload(expand to localstorage/cookie?).)

### CODE REVIEW

**Comment by ZaradarDFDS** [comment]
SSO between browser tabs
When your application is open in multiple tabs and you first sign in the user on one tab, the user is also signed in on the other tabs without being prompted. MSAL.js caches the ID token for the user in the browser localStorage and will sign the user in to the application on the other open tabs.

By default, MSAL.js uses sessionStorage which does not allow the session to be shared between tabs. To get SSO between tabs, make sure to set the cacheLocation in MSAL.js to localStorage as shown below.

JavaScript

Kopiraj
const config = {
    auth: {
        clientId: "abcd-ef12-gh34-ikkl-ashdjhlhsdg"
    },
    cache: {
        cacheLocation: 'localStorage'
    }
}

const myMSALObj = new UserAgentApplication(config);

**Comment by ZaradarDFDS** [comment]
Det kan være vi lige skal teste MSAL med localStorage, istedet for session storage ift. nogle af de caching issues du oplever. :)

File: src/Blaster.WebApi/Features/Capabilities/capabilityservice.js [hunk 1]
```diff
@@ -11,8 +11,8 @@ export default class CapabilityService {
     }
 
     getAll() {
-        return this.client.get(this.baseUrl)
-            .then(data => data.items || []);
+        var x = this.client.get(this.baseUrl);
```
Comment (ZaradarDFDS) [inline src/Blaster.WebApi/Features/Capabilities/capabilityservice.js]: Use descriptive variable names. Even thou the X-men are kind cool ;)

File: src/Blaster.WebApi/Features/Capabilities/main.js [hunk 2]
```diff
@@ -123,7 +123,12 @@ const app = new Vue({
     },
     mounted: function () {
         jq.ready
-            .then(() => capabilityService.getAll())
+            .then(() => {
+                var x = capabilityService.getAll();
```
Comment (ZaradarDFDS) [inline src/Blaster.WebApi/Features/Capabilities/main.js]: Variable naming :)

File: src/Blaster.WebApi/Features/Shared/httpclient.js [hunk 3]
```diff
@@ -1,6 +1,32 @@
 import jq from "jquery";
 
 const baseUrl = `${window.basePath}`;
+const axios = require('axios').default.create();
+
+axios.interceptors.request.use(
+    req => {
+        const internalHttpClient = new HttpClient();
+        console.log(req.url + " has been hit");
```
Comment (ZaradarDFDS) [inline src/Blaster.WebApi/Features/Shared/httpclient.js]: Do we use UglifyJsPlugin in webpack to remove console.log statements from production env?

File: src/Blaster.WebApi/Features/Capabilities/capabilityservice.js (hunk 1 above)
Comment (SEQUOIIA) [reply src/Blaster.WebApi/Features/Capabilities/capabilityservice.js]: Missed it, thanks!

File: src/Blaster.WebApi/Features/Capabilities/main.js (hunk 2 above)
Comment (SEQUOIIA) [reply src/Blaster.WebApi/Features/Capabilities/main.js]: Missed it, thanks!

File: src/Blaster.WebApi/Features/Shared/httpclient.js (hunk 3 above)
Comment (SEQUOIIA) [reply src/Blaster.WebApi/Features/Shared/httpclient.js]: Nope, will be removed, merely just for visual feedback when demonstrating/trying new things.

**Comment by SEQUOIIA** [comment]
> 
> 
> SSO between browser tabs
> When your application is open in multiple tabs and you first sign in the user on one tab, the user is also signed in on the other tabs without being prompted. MSAL.js caches the ID token for the user in the browser localStorage and will sign the user in to the application on the other open tabs.
> 
> By default, MSAL.js uses sessionStorage which does not allow the session to be shared between tabs. To get SSO between tabs, make sure to set the cacheLocation in MSAL.js to localStorage as shown below:
> 
> const config = {
> auth: {
> clientId: "abcd-ef12-gh34-ikkl-ashdjhlhsdg"
> },
> cache: {
> cacheLocation: 'localStorage'
> }
> }

I've attempted to use both sessionStorage and localStorage(https://github.com/dfds/blaster/pull/134/commits/9c68b9299113bdb2c585a845f681ccd11eda929c#diff-d74c6eefb9754a694fc48ac43ee43e38) without much luck, which lead to my confusion whether it's an issue with the library or just me doing something that wasn't intended.

File: src/Blaster.WebApi/Features/Shared/httpclient.js (hunk 3 above)
Comment (ZaradarDFDS) [reply src/Blaster.WebApi/Features/Shared/httpclient.js]: Maybe we should add the uglify plugin to sleep better at night? :D

Comment (SEQUOIIA) [reply src/Blaster.WebApi/Features/Shared/httpclient.js]: Sounds reasonable! :)

**Comment by ZaradarDFDS** [comment]
Har du prøvet uden en pop up? Ved ikke lige helt hvordan page session on end virker med popup + msal, så måske er det værd at overveje?

**Comment by SEQUOIIA** [comment]
> Har du prøvet uden en pop up? Ved ikke lige helt hvordan page session on end virker med popup + msal, så måske er det værd at overveje?

Aye, der er ingen popup efter jeg satte redirectUri som du nævnte. Men har samme issue både med og uden.

**Comment by SEQUOIIA** [comment]
Oh lord, looks like my IDE(Rider) didn't like some files in 98cf8a8053b237d5fae6326afc1c32b820cdf7cc, I even checked diff in vscode before commiting, but it didn't show that it would actually replace most of the file.

**Comment by SEQUOIIA** [comment]
Trying to hook it up with Capability-service results in ' invalid_token - The signature is invalid' response, investigating.

**Comment by ZaradarDFDS** [comment]
> Trying to hook it up with Capability-service results in ' invalid_token - The signature is invalid' response, investigating

I assume this is no longer relevant. :)

**Comment by SEQUOIIA** [comment]
> > Trying to hook it up with Capability-service results in ' invalid_token - The signature is invalid' response, investigating
> 
> I assume this is no longer relevant. :)

Yup, that is indeed what we fixed yesterday.

**Pull request merged**